sw/device/lib/dif/dif_keymgr.h

```diff
@@ -0,0 +1,648 @@
+ * is used to compute the secret value for the next state of key manager and
+ * initiates a transition to the next state, i.e. Initialized -> CreatorRootKey
+ * -> OwnerIntermediateKey -> OwnerRootKey -> Disabled. State transitions are
+ *  irreversible during a power cycle. Key manager must be initialized using
+ *  `dif_keymgr_initialize_state()` prior to calling this function.
+ *
+ * This is an asynchronous function because key manager state transitions
+ * involve KMAC operations that can take some time to complete. Clients must
+ * check the status of key manager using `dif_keymgr_get_status()` before
+ * calling other functions in this library.
+ *
+ * This function also checks that key manager is in the correct state for the
+ * expected next state.
+ *
+ * @param keymgr A key manager handle.
+ * @param params State transition parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_lockable_result_t dif_keymgr_advance_state(
+    const dif_keymgr_t *keymgr, dif_keymgr_advance_state_params_t params);
+
+/**
+ * Disables key manager.
+ *
+ * This function disables key manager until the next power cycle by making it
+ * transition to Disabled state. Disabled state is a terminal state where key
+ * manager is no longer operational and its secret value is a random value.
+ *
+ * @param keymgr A key manager handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_lockable_result_t dif_keymgr_disable(const dif_keymgr_t *keymgr);
+
+/**
+ * Status code bit flags.
+ *
+ * See also: `dif_keymgr_status_codes_t`.
+ */
+typedef enum dif_keymgr_status_code {
+  /**
+   * Software invoked an invalid operation.
+   */
+  kDifKeymgrStatusCodeInvalidOperation = (1 << 0),
+  /**
+   * Software provided an invalid input.
+   */
+  kDifKeymgrStatusInvalidInput = (1 << 1),
+  /**
+   * Key manager issued invalid data to KMAC interface.
+   */
+  kDifKeymgrStatusInvalidKMACInput = (1 << 2),
+  /**
+   * KMAC returned an invalid output.
+   */
+  kDifKeymgrStatusInvalidKMACOutput = (1 << 3),
+  /**
+   * Key manager is idle.
+   */
+  kDifKeymgrStatusCodeIdle = (1 << 4),
+} dif_keymgr_status_code_t;
+
+/**
+ * A set of status codes.
+ *
+ * See also: `dif_keymgr_status_code_t`.
+ */
+typedef uint8_t dif_keymgr_status_codes_t;
+
+/**
+ * Key manager status.
+ */
+typedef struct dif_keymgr_status {
+  /**
+   * Current state of key manager.
+   */
+  dif_keymgr_state_t state;
+  /**
+   * Operational status of key manager.
+   *
+   * The following snippet can be used to check if key manager is idle:
+   *
+   *   bool is_idle = (status.status_codes & kDifKeymgrStatusCodeIdle);
+   *
+   * The following snippet can be used to check if key manager is idle and
+   * error-free:
+   *
+   *   bool is_idle_and_ok = (status.status_codes == kDifKeymgrStatusCodeIdle);
+   */
+  uint32_t status_codes;
+} dif_keymgr_status_t;
```

Comment:
Do these *need* to be queried together?

Reply:
Not really, but currently HW handles initialization in a different way for the sake of efficiency (not as an operation but as a separate command and reports its status in WORKING_STATE.STATE instead of OP_STATUS.STATUS). I tried to abstract this away (HW may change to do the same). Getting the status would require reading both registers anyway so I thought it would be convenient to keep everything related in a single struct.